### PR TITLE
OS signal handling

### PR DESCRIPTION
**Pull request by mdsjip**

### CODE REVIEW

File: astro/exec2/process.go [hunk 1]
```diff
@@ -106,18 +107,65 @@ func (p *Process) Run() error {
...
+					logger.Error.Print("error sending signal", sig, err)
+				}
+			case chErr := <-waitCh:
+				// Record run time
+				resultTime = time.Since(started)
+
+				// Subprocess exited. Get the return code, if we can
+				var waitStatus syscall.WaitStatus
+				logger.Trace.Printf("exec2: command exit code from Process: %v\n", p.ExitCode())
+				if exitError, ok := chErr.(*exec.ExitError); ok {
```
Comment (dansimau) [inline astro/exec2/process.go]: Isn't this just reimplementing `p.ExitCode()`?

File: astro/exec2/process.go [hunk 2]
```diff
@@ -106,18 +107,65 @@ func (p *Process) Run() error {
...
 	}
 
+	var resultError error
+	var resultTime time.Duration
+
 	// Run the process
 	started := time.Now()
-	err := p.execCmd.Run()
+	if err := p.execCmd.Start(); err != nil {
+		logger.Error.Print(err) // Command not found on PATH, not executable, etc.
```
Comment (dansimau) [inline astro/exec2/process.go]: Let's get rid of all `logger.Error` statements. If there are errors, we should bubble them up to the caller to decide what to do. Anything other than that is for tracing/debugging.

File: astro/utils/parallel.go [hunk 3]
```diff
@@ -17,21 +17,41 @@
...
-		wg.Add(1)
-		go func(fn func()) {
-			defer wg.Done()
-			fn()
-			<-guard
-		}(fn)
+		// https://stackoverflow.com/a/25306241
+		guard <- struct{}{} // would block if guard channel is already filled
+		if isShutDown {
+			fmt.Println("Shutting down. No new terraform processes will be launched.")
```
Comment (dansimau) [inline astro/utils/parallel.go]: This function is a utility function and has nothing to do with Terraform.

Do you think it's an acceptable design to make `Parallel` stop executing fns on SIGTERM? It kinda feels like Parallel should be more general than that.

What happened to the idea of return a context allowing the caller can cancel it?

File: astro/exec2/process.go (hunk 2 above)
Comment (mdsjip) [reply astro/exec2/process.go]: I've changed this to accumulate errors using multierror

File: astro/utils/parallel.go (hunk 3 above)
Comment (mdsjip) [reply astro/utils/parallel.go]: Now it receives a context with cancel

File: astro/exec2/process.go (hunk 1 above)
Comment (mdsjip) [reply astro/exec2/process.go]: Yes, it was for local debugging, removed this part

**Comment by dansimau** [comment]
Superseded by #49.

**Pull request closed without merging**